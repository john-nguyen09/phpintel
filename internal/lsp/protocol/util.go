package protocol

import (
	"strconv"
	"strings"
)

// IsInRange compare whether a position is within a range
func IsInRange(position Position, theRange Range) int {
	start := theRange.Start
	end := theRange.End

	if position.Line < start.Line || (position.Line == start.Line && position.Character < start.Character) {
		return -1
	}
	if position.Line > end.Line || (position.Line == end.Line && position.Character >= end.Character) {
		if ComparePos(start, end) == 0 {
			return -1
		}
		return 1
	}
	return 0
}

// ComparePos returns -1, 0 or 1 indicating whether a is before, equal to or after b
func ComparePos(a Position, b Position) int {
	if a.Line < b.Line {
		return -1
	}
	if a.Line == b.Line {
		if a.Character < b.Character {
			return -1
		}
		if a.Character == b.Character {
			return 0
		}
		if a.Character > b.Character {
			return 1
		}
	}
	return 1
}

// CompareRange compares 2 ranges
func CompareRange(a Range, b Range) int {
	if IsInRange(a.End, b) < 0 {
		return -1
	}
	if IsInRange(a.Start, b) > 0 {
		return 1
	}
	return 0
}

// PosFromString creates Position from string
func PosFromString(str string) Position {
	values := strings.Split(str, ":")
	line, err := strconv.Atoi(values[0])
	if err != nil {
		panic(err)
	}
	character, err := strconv.Atoi(values[1])
	if err != nil {
		panic(err)
	}
	return Position{
		Line:      line,
		Character: character,
	}
}

// RangeFromString creates Range from string
func RangeFromString(str string) Range {
	values := strings.Split(str, "-")
	start := PosFromString(values[0])
	end := PosFromString(values[1])
	return Range{
		Start: start,
		End:   end,
	}
}
