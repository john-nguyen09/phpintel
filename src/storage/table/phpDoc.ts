import { DbStore, LevelDatasource, SubStore } from "../db";
import { Serializer, Deserializer } from "../serializer";
import { injectable } from "inversify";
import { PhpDocument } from "../../symbol/phpDocument";
import { ImportTable } from "../../type/importTable";
import * as AsyncLock from "async-lock";

@injectable()
export class PhpDocumentTable {
    private static lock = new AsyncLock();

    private db: DbStore;

    constructor(level: LevelDatasource) {
        this.db = new SubStore(level, {
            name: 'php_document',
            version: 1,
            valueEncoding: PhpDocEncoding
        });
    }

    async put(phpDoc: PhpDocument) {
        return this.db.put(phpDoc.uri, phpDoc);
    }

    async get(uri: string): Promise<PhpDocument | null> {
        try {
            return await this.db.get(uri);
        } catch {
            return null;
        }
    }

    async remove(uri: string) {
        return this.db.del(uri);
    }

    async getAllStream<T>(callback: (phpDoc: PhpDocument) => void) {
        const db = this.db;

        return await new Promise<void>((resolve, reject) => {
            db.createReadStream()
                .on('data', (data) => {
                    callback(data.value);
                })
                .on('end', () => {
                    resolve();
                })
                .on('error', (err) => {
                    if (err) {
                        reject(err);
                    }
                });
        });
    }

    static async acquireLock(uri: string, action: () => void | PromiseLike<void>): Promise<void> {
        return PhpDocumentTable.lock.acquire(uri, action);
    }
}

export const PhpDocEncoding = {
    type: 'php-doc-encoding',
    encode(phpDoc: PhpDocument): Buffer {
        let serializer = new Serializer();

        serializer.setString(phpDoc.uri);
        serializer.setString(phpDoc.text);
        serializer.setInt32(phpDoc.modifiedTime);

        serializer.setNamespaceName(phpDoc.importTable.namespace);

        let keys = Object.keys(phpDoc.importTable.imports);
        serializer.setInt32(keys.length);
        for (let key of keys) {
            serializer.setString(key);
            serializer.setString(phpDoc.importTable.imports[key]);
        }

        return serializer.getBuffer();
    },
    decode(buffer: Buffer): PhpDocument {
        let deserializer = new Deserializer(buffer);
        let phpDoc = new PhpDocument(deserializer.readString(), deserializer.readString());
        phpDoc.modifiedTime = deserializer.readInt32();

        let importTable = new ImportTable();
        importTable.namespace = deserializer.readNamespaceName();
        let numOfKeys = deserializer.readInt32();

        for (let i = 0; i < numOfKeys; i++) {
            importTable.imports[deserializer.readString()] = deserializer.readString();
        }

        phpDoc.importTable = importTable;

        return phpDoc;
    },
    buffer: true
} as Level.Encoding;