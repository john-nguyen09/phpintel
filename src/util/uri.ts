import URI from 'vscode-uri/lib/umd';
import * as path from 'path';

export function pathToUri(filePath: string): string {
    filePath = path.resolve(filePath).replace(/\\/g, '/');

    if (filePath[0] != '/') {
        filePath = '/' + filePath;
    }
    
    return encodeURI('file://' + filePath);
}

export function uriToPath(uri: string) {
    return URI.parse(uri).fsPath;
}

export function toRelative(uri: string) {
    let baseUri = pathToUri(path.resolve(__dirname, '..' , '..'));

    if (uri.indexOf(baseUri) === 0) {
        return uri.substr(baseUri.length + 1);
    }

    return uri;
}