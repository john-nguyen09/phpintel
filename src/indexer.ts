import * as fs from "fs";
import * as path from "path";
import { promisify } from "util";
import { PhpFile } from "./phpFile";
import { TreeAnalyser } from "./treeAnalyser";

const readdirAsync = promisify(fs.readdir);
const statAsync = promisify(fs.stat);

export interface FileInfo {
    filePath: fs.PathLike;
    stats: fs.Stats;
}

export namespace FileInfo {
    export async function create(path: fs.PathLike): Promise<FileInfo> {
        return {
            filePath: path,
            stats: await statAsync(path)
        }
    }
}

export namespace Indexer {
    export async function indexWorkspace(directory: fs.PathLike): Promise<void> {
        let directories: fs.PathLike[] = [
            directory
        ];
        const promises: Promise<void>[] = [];

        while (directories.length > 0) {
            let dir = directories.shift();
            if (dir === undefined) {
                continue;
            }
            let files = await readdirAsync(dir);

            for (let file of files) {
                const fileInfo = await FileInfo.create(path.join(dir.toString(), file));

                if (fileInfo.stats.isDirectory()) {
                    directories.push(fileInfo.filePath);

                    continue;
                }

                if (file.endsWith('.php')) {
                    promises.push(PhpFile.create(fileInfo).then((phpFile) => {
                        TreeAnalyser.analyse(phpFile);
                    }));
                }
            }
        }

        await Promise.all(promises);
    }
}