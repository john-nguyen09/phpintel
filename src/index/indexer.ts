import * as fs from "fs";
import * as path from "path";
import { promisify } from "util";
import { pathToUri, uriToPath } from "../util/uri";
import { SymbolParser } from "../symbol/symbolParser";
import { PhpDocument } from "../symbol/phpDocument";
import { injectable } from "inversify";
import { Traverser } from "../traverser";
import { ClassTable } from "../storage/table/class";
import { ClassConstantTable } from "../storage/table/classConstant";
import { ConstantTable } from "../storage/table/constant";
import { FunctionTable } from "../storage/table/function";
import { MethodTable } from "../storage/table/method";
import { PropertyTable } from "../storage/table/property";
import { PhpDocumentTable } from "../storage/table/phpDoc";
import { GlobalVariableTable } from "../storage/table/globalVariable";

const readdirAsync = promisify(fs.readdir);
const readFileAsync = promisify(fs.readFile);
const statAsync = promisify(fs.stat);

export interface PhpFileInfo {
    filePath: string;
    fstats: fs.Stats;
}

export namespace PhpFileInfo {
    export async function createFileInfo(filePath: string): Promise<PhpFileInfo> {
        return {
            filePath: filePath,
            fstats: await statAsync(filePath)
        };
    }
}

@injectable()
export class Indexer {
    private openedUri: Set<string> = new Set<string>();

    constructor(
        private treeTraverser: Traverser,
        private phpDocTable: PhpDocumentTable,
        private classTable: ClassTable,
        private classConstantTable: ClassConstantTable,
        private constantTable: ConstantTable,
        private functionTable: FunctionTable,
        private methodTable: MethodTable,
        private propertyTable: PropertyTable,
        private globalVariableTable: GlobalVariableTable
    ) { }

    async getOrCreatePhpDoc(uri: string): Promise<PhpDocument> {
        let phpDoc = await this.phpDocTable.get(uri);

        if (phpDoc === null) {
            let fileContent = (await readFileAsync(uriToPath(uri))).toString('utf-8');
            phpDoc = new PhpDocument(uri, fileContent);
        }

        return phpDoc;
    }

    async syncFileSystem(fileInfo: PhpFileInfo): Promise<void> {
        let fileUri = pathToUri(fileInfo.filePath);
        const fileModifiedTime = Math.round(fileInfo.fstats.mtime.getTime() / 1000);

        let phpDoc = await this.getOrCreatePhpDoc(fileUri);
        if (phpDoc.modifiedTime !== fileModifiedTime) {
            phpDoc.modifiedTime = fileModifiedTime;
            await this.indexFile(phpDoc);
        }
    }

    async indexFile(phpDoc: PhpDocument): Promise<void> {
        let symbolParser = new SymbolParser(phpDoc);

        this.treeTraverser.traverse(phpDoc.getTree(), [symbolParser]);
        await this.indexPhpDocument(symbolParser.getPhpDoc());
    }

    async indexWorkspace(directory: string): Promise<void> {
        let directories: string[] = [
            directory
        ];
        const promises: Promise<void>[] = [];

        while (directories.length > 0) {
            let dir = directories.shift();
            if (dir === undefined) {
                continue;
            }
            let files = await readdirAsync(dir);

            for (let file of files) {
                const fileInfo = await PhpFileInfo.createFileInfo(path.join(dir, file));

                if (file.endsWith('.php')) {
                    promises.push(this.syncFileSystem(fileInfo));
                } else if (fileInfo.fstats.isDirectory()) {
                    directories.push(fileInfo.filePath);
                }
            }
        }

        await Promise.all(promises);
    }

    public async open(uri: string): Promise<void> {
        this.openedUri.add(uri);
        const phpDoc = await this.getOrCreatePhpDoc(uri);
        await this.indexFile(phpDoc);
    }

    public close(uri: string): void {
        this.openedUri.delete(uri);
    }

    private async removeSymbolsByDoc(uri: string) {
        return Promise.all([
            this.classTable.removeByDoc(uri),
            this.classConstantTable.removeByDoc(uri),
            this.constantTable.removeByDoc(uri),
            this.functionTable.removeByDoc(uri),
            this.methodTable.removeByDoc(uri),
            this.propertyTable.removeByDoc(uri),
            this.globalVariableTable.removeByDoc(uri),
        ]);
    }

    private async indexPhpDocument(doc: PhpDocument): Promise<void> {
        await this.removeSymbolsByDoc(doc.uri);

        const promises: Promise<void | void[] | void[][]>[] = [];

        for (const theClass of doc.classes) {
            promises.push(this.classTable.put(doc, theClass));
        }

        for (const classConstant of doc.classConstants) {
            promises.push(this.classConstantTable.put(doc, classConstant));
        }

        for (const constant of doc.constants) {
            promises.push(this.constantTable.put(doc, constant));
        }

        for (const func of doc.functions) {
            promises.push(this.functionTable.put(doc, func));
        }

        for (const method of doc.methods) {
            promises.push(this.methodTable.put(doc, method));
        }

        for (const property of doc.properties) {
            promises.push(this.propertyTable.put(doc, property));
        }

        for (const globalVariable of doc.globalVariables) {
            globalVariable.assignExtraTypeForVariables();
            promises.push(this.globalVariableTable.put(doc, globalVariable));
        }

        await Promise.all(promises);
        await this.phpDocTable.put(doc, this.openedUri.has(doc.uri));
    }
}