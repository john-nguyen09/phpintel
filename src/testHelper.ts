import { PhpDocument } from "./symbol/phpDocument";
import { pathToUri } from "./util/uri";
import * as fs from "fs";
import { SymbolParser } from "./symbol/symbolParser";
import { Traverser } from "./traverser";
import { TreeNode } from "./util/parseTree";
import { Parser, Phrase, phraseTypeToString, tokenTypeToString } from "php7parser";
import * as path from "path";
import { inspect } from "util";

export function indexFiles(filePaths: string[]): PhpDocument[] {
    let phpDocs: PhpDocument[] = [];
    let treeTraverser = new Traverser();

    for (let filePath of filePaths) {
        const fileUri = pathToUri(filePath);
        const fileContent = fs.readFileSync(filePath).toString();
        const symbolParser = new SymbolParser(new PhpDocument(fileUri, fileContent));
        const parseTree = Parser.parse(fileContent);
        
        treeTraverser.traverse(parseTree, [symbolParser]);

        phpDocs.push(symbolParser.getTree());
    }

    return phpDocs;
}

export function getCaseDir(): string {
    return path.resolve(__dirname, "..", "case");
}

export function getDebugDir(): string {
    return path.resolve(__dirname, '..', 'debug');
}

export function dumpToDebug(name: string, object: any, depth?: number): void {
    fs.writeFile(
        path.resolve(__dirname, '..', 'debug', name),
        inspect(object, {
            depth: depth
        }),
        (err) => {
            console.error(err);
        }
    );
}

export function dumpAstToDebug(name: string, parseTree: Phrase): void {
    fs.writeFile(
        path.resolve(__dirname, '..', 'debug', name),
        JSON.stringify(parseTree, (key, value) => {
            if (key == 'modeStack') {
                return undefined;
            }

            if (key == 'phraseType') {
                return phraseTypeToString(value);
            }

            if (key == 'tokenType') {
                return tokenTypeToString(value);
            }

            return value;
        }, 2),
        (err) => {
            if (err) {
                console.log(err);
            }
        }
    );
}