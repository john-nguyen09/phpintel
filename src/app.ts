import "reflect-metadata";
import { LevelDatasource } from "./storage/db";
import { LogWriter } from "./service/logWriter";
import { Hasher } from "./service/hasher";
import { Container, interfaces } from "inversify";
import { Indexer } from "./index/indexer";
import { ClassTable } from "./storage/table/class";
import { ClassConstantTable } from "./storage/table/classConstant";
import { ConstantTable } from "./storage/table/constant";
import { FunctionTable } from "./storage/table/function";
import { MethodTable } from "./storage/table/method";
import { PropertyTable } from "./storage/table/property";
import { Traverser } from "./traverser";
import { ReferenceTable } from "./storage/table/reference";
import { PhpDocumentTable } from "./storage/table/phpDoc";
import { IConnection } from "vscode-languageserver";
import { ScopeVarTable } from "./storage/table/scopeVar";

export interface AppOptions {
    storage: string;
}

export class Application {
    protected container: Container = new Container();
    protected options: AppOptions = {
        storage: ''
    };

    constructor(storageLocation: string, connection?: IConnection) {
        this.container.bind<IConnection | undefined>('IConnection').toConstantValue(connection);

        this.initBind();
        this.initStorage(storageLocation);
    }

    public getContainer(): Container {
        return this.container;
    }

    public async clearCache() {
        const db = this.container.get<LevelDatasource>(LevelDatasource).getDb();
        let promises: Promise<void>[] = [];

        return new Promise<void>((resolve, reject) => {
            db.createKeyStream()
                .on('data', (key) => {
                    promises.push(db.del(key));
                })
                .on('end', () => {
                    Promise.all(promises).then(() => { resolve() });
                })
                .on('error', (err: Error) => {
                    reject(err);
                });
        });
    }

    public async shutdown() {
        await this.container.get<LevelDatasource>(LevelDatasource).getDb().close();
    }

    protected initStorage(location: string) {
        this.options.storage = location;

        let datasource = new LevelDatasource(location);
        this.container.bind<LevelDatasource>(LevelDatasource).toConstantValue(datasource);

        // Tables
        this.container.bind<ClassTable>(ClassTable).toSelf().inSingletonScope();
        this.container.bind<ClassConstantTable>(ClassConstantTable).toSelf().inSingletonScope();
        this.container.bind<ConstantTable>(ConstantTable).toSelf().inSingletonScope();
        this.container.bind<FunctionTable>(FunctionTable).toSelf().inSingletonScope();
        this.container.bind<MethodTable>(MethodTable).toSelf().inSingletonScope();
        this.container.bind<PropertyTable>(PropertyTable).toSelf().inSingletonScope();
        this.container.bind<PhpDocumentTable>(PhpDocumentTable).toSelf().inSingletonScope();
        this.container.bind<ReferenceTable>(ReferenceTable).toSelf().inSingletonScope();
        this.container.bind<ScopeVarTable>(ScopeVarTable).toSelf().inSingletonScope();
    }

    protected initBind() {
        this.container.bind<LogWriter>(LogWriter).toSelf().inSingletonScope();
        this.container.bind<Hasher>(Hasher).toSelf().inSingletonScope();
        this.container.bind<Indexer>(Indexer).toSelf().inSingletonScope();
        this.container.bind<Traverser>(Traverser).toSelf().inSingletonScope();
    }
}

export namespace App {
    let app: Application;

    export function init(storageLocation: string, connection?: IConnection) {
        app = new Application(storageLocation, connection);
    }

    export function get<T>(
        identifier: string | symbol | interfaces.Newable<T> | interfaces.Abstract<T>
    ): T {
        return app.getContainer().get(identifier);
    }

    export function set<T>(
        identifier: string,
        value: T
    ): void {
        app.getContainer().bind<T>(identifier).toConstantValue(value);
    }

    export async function clearCache() {
        return await app.clearCache();
    }

    export async function shutdown() {
        return await app.shutdown();
    }
}
