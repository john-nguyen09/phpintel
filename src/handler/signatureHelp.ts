import { TextDocumentPositionParams, SignatureHelp } from "vscode-languageserver";
import { App } from "../app";
import { PhpDocumentTable } from "../storage/table/phpDoc";
import { ReferenceTable } from "../storage/table/reference";
import { RefResolver } from "./refResolver";
import { Reference } from "../symbol/reference";

export namespace SignatureHelpProvider {
    export async function provide(params: TextDocumentPositionParams): Promise<SignatureHelp | null> {
        const phpDocTable = App.get<PhpDocumentTable>(PhpDocumentTable);
        const refTable = App.get<ReferenceTable>(ReferenceTable);
        let signatureHelp: SignatureHelp | null = null;


        await PhpDocumentTable.acquireLock(params.textDocument.uri, async () => {
            const phpDoc = await phpDocTable.get(params.textDocument.uri);

            if (phpDoc === null) {
                return;
            }

            const offset = phpDoc.getOffset(params.position.line, params.position.character);
            const ref = await refTable.findAt(phpDoc.uri, offset);

            if (ref === null) {
                return;
            }

            signatureHelp = await RefResolver.getSignatureHelp(phpDoc, ref, offset);
        });

        return signatureHelp;
    }
}