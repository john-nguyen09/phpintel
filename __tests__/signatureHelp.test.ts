import { App } from "../src/app";
import * as path from "path";
import { getDebugDir, getCaseDir, dumpAstToDebug } from "../src/testHelper";
import { Indexer, PhpFileInfo } from "../src/index/indexer";
import { PhpDocumentTable } from "../src/storage/table/phpDoc";
import { pathToUri } from "../src/util/uri";
import { SignatureHelpProvider } from "../src/handler/signatureHelp";

interface SignatureHelpTestCase {
    testFile: string;
    line: number;
    character: number;
}

async function testSignatureHelp(definitionFiles: string[], testCases: SignatureHelpTestCase[]) {
    const indexer = App.get<Indexer>(Indexer);
    const phpDocTable = App.get<PhpDocumentTable>(PhpDocumentTable);

    for (let definitionFile of definitionFiles) {
        await indexer.syncFileSystem(
            await PhpFileInfo.createFileInfo(path.join(getCaseDir(), definitionFile))
        );
    }

    for (const testCase of testCases) {
        const testFileUri = pathToUri(testCase.testFile);

        await indexer.syncFileSystem(await PhpFileInfo.createFileInfo(testCase.testFile));

        const phpDoc = await phpDocTable.get(testFileUri);

        if (phpDoc === null) {
            return;
        }

        const signatureHelp = await SignatureHelpProvider.provide({
            position: { line: testCase.line, character: testCase.character },
            textDocument: {
                uri: testFileUri
            }
        });

        expect(signatureHelp).toMatchSnapshot();
    }
}

beforeAll(() => {
    App.init(path.join(getDebugDir(), 'storage'));
});

beforeEach(async () => {
    await App.clearCache();
});

afterAll(async () => {
    await App.shutdown();
});

describe('provide signature help', () => {
    it('shows list of parameters', async () => {
        await testSignatureHelp([
            'function_declare.php',
        ], [
            { testFile: path.join(getCaseDir(), 'signatureHelp', 'function.php'), line: 2, character: 16 },
            { testFile: path.join(getCaseDir(), 'signatureHelp', 'function.php'), line: 2, character: 21 },
            { testFile: path.join(getCaseDir(), 'signatureHelp', 'method.php'), line: 4, character: 23 },
            { testFile: path.join(getCaseDir(), 'signatureHelp', 'method.php'), line: 12, character: 26 },
        ]);
    });

    it('parameters for type designator', async () => {
        await testSignatureHelp([
            'class_methods.php',
        ], [
            { testFile: path.join(getCaseDir(), 'signatureHelp', 'type_designator.php'), line: 2, character: 20 },
            { testFile: path.join(getCaseDir(), 'signatureHelp', 'type_designator.php'), line: 4, character: 23 },
        ]);
    });
});