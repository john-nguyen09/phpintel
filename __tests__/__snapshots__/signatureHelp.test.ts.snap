// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`provide signature help parameters for type designator 1`] = `
Object {
  "activeParameter": 0,
  "activeSignature": 0,
  "signatures": Array [
    Object {
      "documentation": "",
      "label": "function __construct(null $optional, null $optional2)",
      "parameters": Array [
        Object {
          "documentation": "",
          "label": "$optional",
        },
        Object {
          "documentation": "",
          "label": "$optional2",
        },
      ],
    },
  ],
}
`;

exports[`provide signature help parameters for type designator 2`] = `
Object {
  "activeParameter": 1,
  "activeSignature": 0,
  "signatures": Array [
    Object {
      "documentation": "",
      "label": "function __construct(null $optional, null $optional2)",
      "parameters": Array [
        Object {
          "documentation": "",
          "label": "$optional",
        },
        Object {
          "documentation": "",
          "label": "$optional2",
        },
      ],
    },
  ],
}
`;

exports[`provide signature help shows list of parameters 1`] = `
Object {
  "activeParameter": 0,
  "activeSignature": 0,
  "signatures": Array [
    Object {
      "documentation": "",
      "label": "function \\\\test_function1(string $stringParam, int $intParam, boolean $boolParam, float $floatParam)",
      "parameters": Array [
        Object {
          "documentation": "",
          "label": "$stringParam",
        },
        Object {
          "documentation": "",
          "label": "$intParam",
        },
        Object {
          "documentation": "",
          "label": "$boolParam",
        },
        Object {
          "documentation": "",
          "label": "$floatParam",
        },
      ],
    },
  ],
}
`;

exports[`provide signature help shows list of parameters 2`] = `
Object {
  "activeParameter": 1,
  "activeSignature": 0,
  "signatures": Array [
    Object {
      "documentation": "",
      "label": "function \\\\test_function1(string $stringParam, int $intParam, boolean $boolParam, float $floatParam)",
      "parameters": Array [
        Object {
          "documentation": "",
          "label": "$stringParam",
        },
        Object {
          "documentation": "",
          "label": "$intParam",
        },
        Object {
          "documentation": "",
          "label": "$boolParam",
        },
        Object {
          "documentation": "",
          "label": "$floatParam",
        },
      ],
    },
  ],
}
`;

exports[`provide signature help shows list of parameters 3`] = `
Object {
  "activeParameter": 0,
  "activeSignature": 0,
  "signatures": Array [
    Object {
      "documentation": "",
      "label": "function method1($param1, $param2)",
      "parameters": Array [
        Object {
          "documentation": "",
          "label": "$param1",
        },
        Object {
          "documentation": "",
          "label": "$param2",
        },
      ],
    },
  ],
}
`;

exports[`provide signature help shows list of parameters 4`] = `
Object {
  "activeParameter": 0,
  "activeSignature": 0,
  "signatures": Array [
    Object {
      "documentation": "",
      "label": "function method2($param1, $param2, boolean $param3)",
      "parameters": Array [
        Object {
          "documentation": "",
          "label": "$param1",
        },
        Object {
          "documentation": "",
          "label": "$param2",
        },
        Object {
          "documentation": "",
          "label": "$param3",
        },
      ],
    },
  ],
}
`;
